### PR TITLE
Convert math alignment to single-pass algorithm

### PR DESCRIPTION
**Pull request by SUPERCILEX**

### CODE REVIEW

**Comment by SUPERCILEX** [comment]
Huh, I'm very confused. This passes locally for me.

**Comment by SUPERCILEX** [comment]
Ok, not *quite* single-pass anymore cuz there was a bug, but still vastly superior complexity to the previous algorithm (`O(rows + aligns)` vs `O(rows + aligns * rows)`).

File: library/src/math/align.rs [hunk 1]
```diff
@@ -16,34 +16,29 @@ impl LayoutMath for AlignPointElem {
...
-    let count = rows
-        .iter()
-        .map(|row| {
-            row.iter()
-                .filter(|fragment| matches!(fragment, MathFragment::Align))
-                .count()
-        })
-        .max()
-        .unwrap_or(0);
+    let mut points = Vec::<Abs>::new();
```
Comment (laurmaedje) [inline library/src/math/align.rs:19]: I had a bit of trouble understanding that this vector now holds the widths first, maybe we can rename it to `widths` here and then rebind `let mut points = widths` before the second pass?

Comment (SUPERCILEX) [reply library/src/math/align.rs]: Good point, done.

File: library/src/math/align.rs [hunk 2]
```diff
@@ -16,34 +16,30 @@ impl LayoutMath for AlignPointElem {
...
+                width += fragment.width();
             }
         }
     }
 
+    let mut points = widths;
+    for i in 1..points.len() {
+        let prev = points[i - 1];
+        points[i] += prev;
+    }
```
Comment (owiecc) [inline library/src/math/align.rs:43]: Wouldn't this be done better as a cumsum iterator over widths? 

PS. I don't know Rust but this seems like a common pattern that can be done with the local flavour of folds and scans, avoiding mutability.

Comment (laurmaedje) [reply library/src/math/align.rs:43]: It could, but this way we don't need to pray that the compiler omits the extra allocation it would incur.

**Pull request merged**